### DISCUSSION TITLE
[Inverted FlashList] How to show content on upper side when total content height is less than list height?

### DISCUSSION THREAD

**Discussion by dmahajan980** [Shopify/flash-list#773] [Q&A]
Hello, 

Thanks for this amazing library. We are using this to implement the chat messages list and are hitting some design limitations.

## Desired Behavior

The messages list has two requirements:

1. When the messages are not enough to cover the whole list, show them on top.
2. Otherwise, the chatroom should show the last messages in the first render.

The images below illustrate the above requirements:

| Short Chat | Long Chat  |
|------------|------------|
| Messages should be displayed on top | Last message should be displayed first |
| <img src='https://user-images.githubusercontent.com/37256767/221528766-e49d4ed1-9e74-4e81-8540-627848c6018b.png' height=500 /> | <img src='https://user-images.githubusercontent.com/37256767/221530063-41688891-8757-49ef-b200-52b7f5153fab.png' height=500 /> |

(*Kindly ignore the absence of input box in right-side screenshot. It's taken from the simulator and is a WIP.*)

## Actual Behavior

In the code, the `inverted` prop is enabled to render the list in reverse order and the long-chat case is achieved. However, in the short-chat case, the messages are also rendering on the bottom side.

<img src='https://user-images.githubusercontent.com/37256767/221531314-28f01417-d769-4d14-b3e2-8006deb3dcbc.png' height=500 />

In the desired behavior, the messages should be displayed on top.

## Tried Solutions

So far, we have tried the following:

1. Reversed the array passed to `data` prop, removed `inverted` prop, and set `initialScrollIndex` to the maximum index value (using `data.length - 1`). This didn't work as the list displayed somewhere around middle and not the end. Moreover, the list displayed the top messages first and then scrolled down which isn't very ideal.
2. Calling `scrollToBottom()` method on the list ref inside `useEffect`. This also suffered the same problems we faced in previous point.

## Similar discussion/issue

https://github.com/Shopify/flash-list/discussions/665

## Reproducible

https://snack.expo.dev/@dmahajan98/chat-room-reproducible

---

I think the solution might be along the lines of not using `inverted` prop. However, the solutions that I tried and am aware of could not render the bottom of list correctly. Also, these solutions do not show the bottom in first render but in subscequent renders.

Any help regarding this would be appreciated very much.

Thanks :)

**Comment by lisacher7** [comment]
Hi, did you manage to find a workaround for this?

**Comment by chaesngmin** [comment]
Same issue. Any solution found?

**Comment by iamjoshua** [comment]
Same issue. This should be a simple css matter of applying justifyContent as flex-end (because inverted), but FlashList doesn't honor this container style. I found that a simple conditional switch between FlashList and a standard FlatList when items under some length was a good-enough workaround, but it adds silly complexity to solve something that shouldn't be an issue.

**Comment by nduyvu1511** [comment]
you can try this:
`<View style={{ flexDirection: 'row', justifyContent: 'flex-start' }}>
        <FlashList />
      </View>`

**Reply by nduyvu1511** [reply]
Remember, when wrapping FlashList inside a View with a style like that, adding a new item with a dynamic height (e.g., sending an image message) requires recalculation. Otherwise, FlashList will leave a blank space because FlashList can't receive the height information to calculate the layout.

**Reply by domleboss97** [reply]
@nduyvu1511 This does not work if you want the parent container to take up the full screen

**Reply by domleboss97** [reply]
I need the container around the flashlist to have `flex-grow` set on it, but I want the items of the flashlist aligned to the top within that container. As far as I can tell, there is not a way to achieve this.

**Comment by hongppa324** [comment]
You can put ListHeaderComponent when using inverted FlashList.

**Comment by hongppa324** [comment]
@dmahajan980 Hi. Did you solve this issue?

**Comment by henninghall** [comment]
I got it to work with following patch. 

Not that this is a quick fix that top aligns all inverted lists. We make this better by allowing flexGrow and flexDirection to be set from outer api or create a new additional prop.

@naqvitalha what do you think? I can prepare a PR if you would like this feature? 

This is a limitation that FlatList currently doesn't have. For Flatlist you can set `flexGrow` and `flexDirection` in `contentContainerStyle` and achieve this behaviour. 


```diff
diff --git a/CHANGELOG.md b/CHANGELOG.md
deleted file mode 100644
index 81d937fa0996aa4340eeb790a53d781fe37c4ea0..0000000000000000000000000000000000000000
diff --git a/dist/FlashList.js b/dist/FlashList.js
index 4f33d6bc679ab1f4daafbacf73172dcb08dc0c1f..702b38209fb6fab1af3ae2a760b7e0115dcf3bef 100644
--- a/dist/FlashList.js
+++ b/dist/FlashList.js
@@ -383,7 +383,10 @@ var FlashList = /** @class */ (function (_super) {
                 ? tslib_1.__assign({}, this.getTransform()) : tslib_1.__assign({ flex: 1, overflow: "hidden" }, this.getTransform()) },
             react_1.default.createElement(recyclerlistview_1.ProgressiveListView, tslib_1.__assign({}, restProps, { ref: this.recyclerRef, layoutProvider: this.state.layoutProvider, dataProvider: this.state.dataProvider, rowRenderer: this.emptyRowRenderer, canChangeSize: true, isHorizontal: Boolean(horizontal), scrollViewProps: tslib_1.__assign({ onScrollBeginDrag: this.onScrollBeginDrag, onLayout: this.handleSizeChange, refreshControl: this.props.refreshControl || this.getRefreshControl(), 
                     // Min values are being used to suppress RLV's bounded exception
-                    style: { minHeight: 1, minWidth: 1 }, contentContainerStyle: tslib_1.__assign({ backgroundColor: this.contentStyle.backgroundColor, 
+                    style: { minHeight: 1, minWidth: 1,  }, contentContainerStyle: tslib_1.__assign({ backgroundColor: this.contentStyle.backgroundColor, 
+                        // This patch makes inverted list be top aligned like non-inverted lists.
+                        flexGrow: restProps.inverted ? 1 : undefined,
+                        flexDirection: restProps.inverted ? 'column-reverse' : undefined,
                         // Required to handle a scrollview bug. Check: https://github.com/Shopify/flash-list/pull/187
                         minHeight: 1, minWidth: 1 }, (0, ContentContainerUtils_1.getContentContainerPadding)(this.contentStyle, horizontal)) }, this.props.overrideProps), forceNonDeterministicRendering: true, renderItemContainer: this.itemContainer, renderContentContainer: this.container, onEndReached: this.onEndReached, onEndReachedThresholdRelative: onEndReachedThreshold || undefined, extendedState: this.state.extraData, layoutSize: estimatedListSize, maxRenderAhead: 3 * finalDrawDistance, finalRenderAheadOffset: finalDrawDistance, renderAheadStep: finalDrawDistance, initialRenderIndex: (!this.isInitialScrollIndexInFirstRow() && initialScrollIndex) ||
                     undefined, initialOffset: initialOffset, onItemLayout: this.onItemLayout, onScroll: this.onScroll, onVisibleIndicesChanged: this.viewabilityManager.shouldListenToVisibleIndices
```

**Reply by winterdouglas** [reply]
I applied this fix by using:
```
overrideProps={{contentContainerStyle: styles.contentContainerStyle}}
```
just making sure to respect the internal props and applying your two suggested props, so that did not require a patch.

I was impressed that `column-reverse` didn't actually invert items which comes in handy for this, so that did the trick for now. Applying the same workaround on RN's `FlatList` actually inverts items, so that scares me a tiny bit :D

**Comment by dition0221** [comment]
I fixed it like this

```
<FlashList
  ......
  contentContainerStyle={styles.flashList}
  overrideProps={{ contentContainerStyle: styles.override }}
  inverted
  ......
/>

const styles = StyleSheet.create({
  flashList: {
    padding: 20
  },
  override: {
    flexGrow: 1,
    paddingHorizontal: 20,
    justifyContent: 'flex-end'
  },
});

```

**Reply by vanBlayt** [reply]
@dition0221 tks, solve my problem.

**Reply by siddhant-rl** [reply]
@dition0221 Thank you so much - that solved the issue! You're a lifesaver ❤️

**Reply by namanica** [reply]
I also ended up with this solution, it looks fine! but when i put a dynamic content into the ListFooterComponent sometimes it is not rendered although the place for it is preserved and in logs I can see the changed content (it's a string)

**Comment by zollipaul** [comment]
How does this work with Flastlist v2? Inverted prop is here deprecated. Someone got a working solution?

**Reply by alxlnk** [reply]
https://shopify.github.io/flash-list/docs/v2-migration#step-4-replace-inverted-lists

**Comment by kinthaiofficial** [comment]
The inverted FlashList alignment issue (content floating at bottom when there's less than a full screen's worth) is a known pain point. A few approaches that work:

**Approach 1: `contentContainerStyle` with `flexGrow`**
```tsx
<FlashList
  inverted
  data={messages}
  contentContainerStyle={{ flexGrow: 1, justifyContent: 'flex-end' }}
  estimatedItemSize={60}
/>
```
This tells the scroll container to expand to fill available space, pushing content to the visual top (which is the bottom in inverted mode). Works for many cases.

**Approach 2: `ListFooterComponent` as a spacer**
```tsx
const spacerHeight = Math.max(0, listHeight - contentHeight);

<FlashList
  inverted
  data={messages}
  ListFooterComponent={<View style={{ height: spacerHeight }} />}
  estimatedItemSize={60}
/>
```
The footer in inverted mode renders at the top. Calculating exact spacer height requires knowing the content height, which you can get from `onContentSizeChange`.

**Approach 3: Scroll programmatically on load**
Less elegant, but for chat interfaces specifically: after initial render, if content doesn't fill the list, don't use inverted at all — instead, scroll to the bottom after render with `ref.current?.scrollToEnd()`.

**Root cause**: FlashList (and FlatList with `inverted`) flips the scroll direction with a matrix transform, but `justifyContent` behavior in the scroll container doesn't automatically invert. The content starts at the "bottom" of the container (which visually appears at the top after the transform).

Is this specifically for a chat/messaging interface where you want newest messages at the bottom?
